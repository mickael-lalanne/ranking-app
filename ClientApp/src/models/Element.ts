export interface Element {
    id?: string;
    name: string;
    image: string;
    type?: number;
}

export interface RankedElement {
    elementId: string;
    tierId: string;
    position: number;
}
