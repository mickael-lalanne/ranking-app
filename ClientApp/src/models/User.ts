export type UserId = string | null | undefined;

export interface User {
    id: UserId;
}
