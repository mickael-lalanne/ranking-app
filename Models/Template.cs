namespace ranking_app.Models;

public class TemplateModel
{
    public int Id { get; set; }
    public required string Name { get; set; }
    public required string UserId { get; set; }
    public ICollection<Tier> Tiers { get; set; }
    public ICollection<Element> Elements { get; set; }
    public DateTime CreatedAt { get; set; }
}

public class TemplatePutPayload : TemplateModel
{
    public ICollection<TierPayload> Tiers { get; set; }
    public required ICollection<ElementPayload> Elements { get; set; }
}
