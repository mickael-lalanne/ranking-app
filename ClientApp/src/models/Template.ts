import { Element } from './Element';
import { UserId } from './User';

export interface Template {
    id?: string;
    name: string;
    userId?: UserId;
    tiers: Tier[];
    elements: Element[];
    createdAt?: string; // ex: "2023-11-16T20:26:59.654363Z"
}

export interface Tier {
    id?: string;
    name: string;
    rank: number;
}

export const TIERS_COLORS: string[] = [
    '#FF8C7F', // Rank 1
    '#FFC07D', // Rank 2
    '#FEF782', // Rank 3
    '#7AEA81', // Rank 4
    '#78BDFD', // Rank 5
];

export enum ETemplateMode {
    Viewer = 'viewer',
    Builder = 'builder',
    Editor = 'editor'
};

export enum EEditViewMode {
    EditElement = 'edit-element', // When user wants to edit an element
    EditTier = 'edit-tier', // When user wants to edit a tier
    Hide = 'hide' // When the view isn't displayed or an element
};
