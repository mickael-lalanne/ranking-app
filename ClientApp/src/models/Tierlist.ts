import { RankedElement } from './Element';
import { UserId } from './User';

export interface Tierlist {
    id: string;
    name: string;
    userId: UserId;
    templateId: string;
    rankedElements: RankedElement[];
    createdAt: string; // ex: "2023-11-16T20:26:59.654363Z"
}

export enum ETierlistDragItem {
    Element = 'element'
}
