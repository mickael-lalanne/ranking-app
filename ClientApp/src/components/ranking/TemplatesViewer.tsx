import React, { useEffect, useState } from 'react';
import { TIERS_COLORS, Template, Tier } from '../../models/Template';
import { css } from '@emotion/css';
import IconButton from '@mui/material/IconButton';
import DeleteIcon from '@mui/icons-material/Delete';
import { deleteTemplate, getTemplates } from '../../services/TemplateServices';

const TemplatesViewer = () => {
    const [userTemplates, setUserTemplates] = useState<Template[]>([]);

    // Called when the component is initialized
    useEffect(() => {
        // Get all user templates from the database
        const fetchTemplates: () => Promise<void> = async () => {
            setUserTemplates(await getTemplates());
        };
        fetchTemplates()
            .catch(err => {
                // TODO: handle errors
            });
    }, []);

    /**
     * Called when the user clicked on the "Delete" button
     * Delete the template (both client and server side) 
     */
    // const onDeleteClick = async (templateId: number): Promise<void> => {
    //     await deleteTemplate(templateId);
    //     setUserTemplates(userTemplates.filter(t => t.id !== templateId));
    // };

    /**
     * Some design to display tiers colors in the template preview
     * @param {Tiers} tiers template's tiers
     * @returns {JSX.Element[]} divs containing with the tiers colors in background
     */
    const generateTemplateTiersItems = (tiers: Tier[]): JSX.Element[] => {
        const tierItems: JSX.Element[] = [];

        tiers.forEach(tier => {
            tierItems.push(
                <div className={tier_style} style={{backgroundColor: TIERS_COLORS[tier.rank]}}></div>
            );
        });

        return tierItems;
    };

    /**
     * Display a preview for all user's template
     * @returns {JSX.Element[]} squares with data like the template name and its tiers
     */
    const generateTemplatesPreviewItems = (): JSX.Element[] => {
        const litsItems: JSX.Element[] = [];

        userTemplates.forEach(template => {
            litsItems.push(
                <div className={template_container_style}>
                    <div className="app_spacer"></div>
                    <div>{template.name}</div>
                        {/* <IconButton edge="end" aria-label="delete" onClick={() => onDeleteClick(template.id!)}>
                            <DeleteIcon />
                        </IconButton> */}
                    <div className="app_spacer"></div>
                    {generateTemplateTiersItems(template.tiers)}
                </div>
            );
        });

        return litsItems;
    };

    return (<>
        <div className={templates_container_style}>
            {generateTemplatesPreviewItems()}
        </div>
    </>);
};

export default TemplatesViewer;

/**
 * CSS STYLES
 */
const templates_container_style = css({
    display: 'flex',
    width: 'fit-content'
});

const template_container_style = css({
    display: 'flex',
    alignItems: 'center',
    margin: '25px',
    width: '200px',
    height: '200px',
    flexDirection: 'column',
    justifyContent: 'center',
    boxShadow: 'rgba(149, 157, 165, 0.2) 0px 8px 24px',
    cursor: 'pointer',
    transitionDuration: '500ms',
    ":hover": {
        transform: 'scale(1.15)'
    }
});

const tier_style = css({
    width: '100%',
    height: '5px',
});
