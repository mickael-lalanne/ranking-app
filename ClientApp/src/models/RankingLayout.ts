import { Template } from './Template';
import { Tierlist } from './Tierlist';
import TemplatesViewer from '../components/template/TemplatesViewer';
import TemplateEditor from '../components/template/TemplateEditor';
import { createTemplate, deleteTemplate, updateTemplate } from '../services/TemplateServices';
import TierlistsViewer from '../components/tierlist/TierlistsViewer';
import TierlistsEditor from '../components/tierlist/TierlistsEditor';
import { AppDispatch } from '../app/store';
import { createTierlist, deleteTierlist, updateTierlist } from '../services/TierlistServices';

export type RankingType = Template | Tierlist;

export enum ERankingLayoutMode {
    Viewer = 'viewer',
    Builder = 'builder',
    Editor = 'editor'
};

type EditorComponent = ({ saveHandler, itemToEdit, mode }: EditorComponentProps) => React.JSX.Element;

export interface EditorComponentProps {
    saveHandler: (itemToSave: Template | Tierlist) => Promise<void>;
    itemToEdit?: Template | Tierlist | undefined;
    mode: ERankingLayoutMode;
};

export interface RankingLayoutProps {
    viewerTitle: string;
    viewerSubtitle: string;
    viewerBtnText: string;
    builderTitle: string;
    builderSubtitle: string;
    builderBtnText: string;
    editorTitle: string;
    editorSubtitle: string;
    editorBtnText: string;
    ViewerComponent: ({ editHandler }: {
        editHandler: (template: Template) => void;
    }) => React.JSX.Element;
    EditorComponent: EditorComponent;
    createFunction: (itemToCreate: RankingType, dispatch: AppDispatch) => Promise<any>;
    updateFunction: (itemToUpdate: RankingType, dispatch: AppDispatch) => Promise<any>;
    deleteFunction: (itemIdToDelete: number, dispatch: AppDispatch) => Promise<any>;
}


export const TEMPLATE_LAYOUT_PROPS: RankingLayoutProps = {
    viewerTitle: 'My templates',
    viewerSubtitle: 'Create, Edit or Delete',
    viewerBtnText: 'Add template',
    builderTitle: 'Template creation',
    builderSubtitle: 'Add a name, tiers and elements',
    builderBtnText: 'Cancel',
    editorTitle: 'Template edition',
    editorSubtitle: 'Update your template here',
    editorBtnText: 'Cancel',
    ViewerComponent: TemplatesViewer,
    EditorComponent: TemplateEditor,
    createFunction: createTemplate as (templateToCreate: RankingType, dispatch: AppDispatch) => Promise<Template>,
    updateFunction: updateTemplate as (templateToUpdate: RankingType, dispatch: AppDispatch) => Promise<void>,
    deleteFunction: deleteTemplate
};

export const TIERLIST_LAYOUT_PROPS: RankingLayoutProps = {
    viewerTitle: 'My tierlists',
    viewerSubtitle: 'Create, Edit or Delete',
    viewerBtnText: 'Add tierlist',
    builderTitle: 'Tierlist creation',
    builderSubtitle: 'Rank your elements',
    builderBtnText: 'Cancel',
    editorTitle: 'Tierlist edition',
    editorSubtitle: 'Update your tierlist here',
    editorBtnText: 'Cancel',
    ViewerComponent: TierlistsViewer,
    EditorComponent: TierlistsEditor,
    createFunction: createTierlist as (tierlistToCreate: RankingType) => Promise<Tierlist>,
    updateFunction: updateTierlist as (tierlistToUpdate: RankingType) => Promise<Response>,
    deleteFunction: deleteTierlist
};
