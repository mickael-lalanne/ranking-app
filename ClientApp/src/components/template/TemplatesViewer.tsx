import React from 'react';
import { TIERS_COLORS, Template, Tier } from '../../models/Template';
import { css } from '@emotion/css';
import { useAppSelector } from '../../app/hooks';
import { ViewerComponentProps } from '../../models/RankingLayout';
import { sortedTemplatesSelector } from '../../store/TemplateStore';
import InfoBox from '../shared/InfoBox';

const TemplatesViewer = ({ editHandler } : ViewerComponentProps) => {
    // Retrieve user templates from the store
    const userTemplates: Template[] =
        useAppSelector(state => sortedTemplatesSelector(state.templates));

    /**
     * Some design to display tiers colors in the template preview
     * @param {Tiers} tiers template's tiers
     * @returns {JSX.Element[]} divs containing with the tiers colors in background
     */
    const generateTemplateTiersItems = (tiers: Tier[]): JSX.Element[] => {
        const tierItems: JSX.Element[] = [];

        tiers.forEach(tier => {
            tierItems.push(
                <div className={tier_style} key={tier.id} style={{backgroundColor: TIERS_COLORS[tier.rank]}}></div>
            );
        });

        return tierItems;
    };

    /**
     * Display a preview for all user's template
     * @returns {JSX.Element[]} squares with data like the template name and its tiers
     */
    const generateTemplatesPreviewItems = (): JSX.Element[] => {
        const litsItems: JSX.Element[] = [];

        userTemplates.forEach(template => {
            litsItems.push(
                <div className={template_container_style} key={template.id} onClick={() => editHandler(template)}>
                    <div className="app_spacer"></div>
                    <div style={{ padding: '20px' }}>{template.name}</div>
                    <div className="app_spacer"></div>
                    {generateTemplateTiersItems(template.tiers)}
                </div>
            );
        });

        return litsItems;
    };

    const EmptyMessage = (): React.JSX.Element | undefined => {
        // NO TEMPLATE MESSAGE
        if (userTemplates.length === 0) {
            return <InfoBox content="It seems you don't have any template yet. <br>Click on the button above to start your creation !" />;
        }
    };

    return (<>
        {EmptyMessage()}
        <div className={templates_container_style}>
            {generateTemplatesPreviewItems()}
        </div>
    </>);
};

export default TemplatesViewer;

/**
 * CSS STYLES
 */
const templates_container_style = css({
    display: 'flex',
    flexWrap: 'wrap'
});

const template_container_style = css({
    display: 'flex',
    alignItems: 'center',
    margin: '25px',
    width: '200px',
    height: '200px',
    flexDirection: 'column',
    justifyContent: 'center',
    boxShadow: 'rgba(149, 157, 165, 0.2) 0px 8px 24px',
    cursor: 'pointer',
    transitionDuration: '500ms',
    textAlign: 'center',
    ":hover": {
        transform: 'scale(1.15)'
    }
});

const tier_style = css({
    width: '100%',
    height: '5px',
});
