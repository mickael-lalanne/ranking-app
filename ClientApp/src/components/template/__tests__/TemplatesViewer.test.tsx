import * as React from 'react';
import {
    ExtendedRenderOptions,
    renderWithProviders,
} from '../../../utils/test-utils';
import { screen } from '@testing-library/react';
import TemplatesViewer from '../TemplatesViewer';
import { Template } from '../../../models/Template';
import userEvent from '@testing-library/user-event';

const editHandler: jest.Mock = jest.fn();
const templateSample: Template = {
    name: 'My template name',
    tiers: [{ name: 'My tier name', rank: 0 }],
    elements: [],
};
const storeWithTemplates: ExtendedRenderOptions = {
    preloadedState: {
        application: {
            fetchingTemplates: false,
            loading: false,
            fetchingTierlists: false,
        },
        templates: { templates: [templateSample] },
    },
};

const TemplatesViewerTest = () => {
    return <TemplatesViewer editHandler={editHandler} />;
};

describe('Screenshot', () => {
    it('takes screenshot when fetching templates', () => {
        const component = renderWithProviders(TemplatesViewerTest());

        expect(component.baseElement).toMatchSnapshot('Fetching templates');
    });

    it('takes screenshot without template', () => {
        const component = renderWithProviders(TemplatesViewerTest(), {
            preloadedState: {
                application: {
                    fetchingTemplates: false,
                    loading: false,
                    fetchingTierlists: false,
                },
            },
        });
        expect(component.baseElement).toMatchSnapshot('Without template');
    });

    it('takes screenshot with templates', () => {
        const component = renderWithProviders(
            TemplatesViewerTest(),
            storeWithTemplates
        );
        expect(component.baseElement).toMatchSnapshot('With templates');
    });
});

describe('Common', () => {
    it('clicks on a template', async () => {
        renderWithProviders(TemplatesViewerTest(), storeWithTemplates);

        await userEvent.click(screen.getByText(templateSample.name));
        expect(editHandler).toHaveBeenCalledWith(templateSample);
    });
});
