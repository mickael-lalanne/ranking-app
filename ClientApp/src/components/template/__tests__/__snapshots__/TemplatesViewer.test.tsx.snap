// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Screenshot takes screenshot when fetching templates: Fetching templates 1`] = `
<body>
  <div>
    <div
      class="css-1yhl6zy"
      data-cy="app-loading"
    >
      <div
        class="css-12lhfml"
      >
        <div
          class="css-fvsue4"
        >
          <div
            class="css-qls6c6"
          >
            <div
              class="css-hthcbb"
            >
              <div
                class="css-1b6vs5j css-8q6na4"
              />
              <div
                class="css-1b6vs5j css-0"
              />
              <div
                class="css-1b6vs5j css-1t5wwna"
              />
              <div
                class="css-1b6vs5j css-1gu2kcg"
              />
              <div
                class="css-1b6vs5j css-17zyqrd"
              />
              <div
                class="css-1b6vs5j css-156lazp"
              />
              <div
                class="css-1b6vs5j css-17i4z6a"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="css-xyomv4"
      >
        <div
          class="css-ircun"
        >
          Fetching templates
        </div>
        <div
          class="css-1lfmw4p"
        />
      </div>
    </div>
  </div>
</body>
`;

exports[`Screenshot takes screenshot with templates: With templates 1`] = `
<body>
  <div>
    <div
      class="css-1h8dwix"
    >
      <div
        class="css-kl4npk"
        data-cy="template-preview"
      >
        <div
          class="app_spacer"
        />
        <div
          style="padding: 20px;"
        >
          My template name
        </div>
        <div
          class="app_spacer"
        />
        <div
          class="css-1t6pfkz"
          data-cy="template-tier"
          style="background-color: rgb(255, 140, 127);"
        />
      </div>
    </div>
  </div>
</body>
`;

exports[`Screenshot takes screenshot without template: Without template 1`] = `
<body>
  <div>
    <div
      class="css-afv6yl"
      data-cy="info-box"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1695c0c css-i4bv87-MuiSvgIcon-root"
        data-testid="TipsAndUpdatesIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 20h4c0 1.1-.9 2-2 2s-2-.9-2-2zm-2-1h8v-2H5v2zm11.5-9.5c0 3.82-2.66 5.86-3.77 6.5H5.27c-1.11-.64-3.77-2.68-3.77-6.5C1.5 5.36 4.86 2 9 2s7.5 3.36 7.5 7.5zm4.87-2.13L20 8l1.37.63L22 10l.63-1.37L24 8l-1.37-.63L22 6l-.63 1.37zM19 6l.94-2.06L22 3l-2.06-.94L19 0l-.94 2.06L16 3l2.06.94L19 6z"
        />
      </svg>
      <span>
        It seems you don't have any template yet. 
        <br />
        Click on the button above to start your creation !
      </span>
    </div>
    <div
      class="css-1h8dwix"
    />
  </div>
</body>
`;
